src/DookMulticurveInitializer.sol

```diff
@@ -0,0 +1,442 @@
+        (BalanceDelta balanceDelta, BalanceDelta feesAccrued) = _burn(state.poolKey, positions);
+        balance0 = uint128(balanceDelta.amount0());
+        balance1 = uint128(balanceDelta.amount1());
+        fees0 = uint128(feesAccrued.amount0());
+        fees1 = uint128(feesAccrued.amount1());
+
+        state.poolKey.currency0.transfer(msg.sender, balance0);
+        state.poolKey.currency1.transfer(msg.sender, balance1);
+    }
+
+    /**
+     * @notice Delegates `msg.sender`'s pool authority to another address
+     * @param delegatedAuthority Address to delgate to
+     */
+    function delegateAuthority(address delegatedAuthority) external {
+        getAuthority[msg.sender] = delegatedAuthority;
+    }
+
+    /**
+     * @notice Sets the Doppler hook for a given asset's pool if not already set
+     * @param asset Address to migrate
+     */
+    function setDook(
+        address asset,
+        address dook,
+        bytes calldata onInitializationCalldata,
+        bytes calldata onGraduationCalldata
+    ) external {
+        PoolState memory state = getState[asset];
+        require(state.status == PoolStatus.Locked, WrongPoolStatus(PoolStatus.Locked, state.status));
+
+        (, address timelock,,,,,,,,) = airlock.getAssetData(asset);
+        address authority = getAuthority[timelock];
+        require(msg.sender == authority || msg.sender == timelock, SenderNotAuthorized());
+
+        if (dook != address(0)) require(isDookEnabled[dook], DookNotEnabled());
+        require(state.dook == address(0), DookAlreadySet());
+
+        getState[asset].dook = dook;
+        getState[asset].graduationDookCalldata = onGraduationCalldata;
+        emit SetDook(asset, dook);
+
+        DookMulticurveHook(address(HOOK)).setDook(state.poolKey.toId(), dook);
+        IDook(dook).onInitialization(asset, onInitializationCalldata);
+    }
+
+    /**
+     * @notice Graduates a pool if the conditions are met, this is a one-way operation
+     * @param asset Address of the asset to graduate
+     */
+    function graduate(address asset) external {
+        PoolState memory state = getState[asset];
+        require(state.status == PoolStatus.Locked, WrongPoolStatus(PoolStatus.Locked, state.status));
+
+        address dook = state.dook;
+        if (address(dook) == address(0)) revert CannotMigratePoolNoProvidedDook();
+
+        _canGraduateOrMigrate(state.poolKey.toId(), asset == Currency.unwrap(state.poolKey.currency0), state.farTick);
+
+        getState[asset].status = PoolStatus.Graduated;
+        emit Graduate(asset);
+        IDook(dook).onGraduation(asset, state.graduationDookCalldata);
+    }
+
+    // TODO: I'm really not sure about this pattern as it's a bit risky
+    function updateDynamicFee(address asset, uint24 lpFee) external {
+        PoolState memory state = getState[asset];
+        require(msg.sender == state.dook, SenderNotAuthorized());
+        poolManager.updateDynamicLPFee(getState[asset].poolKey, lpFee);
+    }
```

Comment:
Note that we can only update the dynamic fee if we set the pool fee as `DYNAMIC_FEE_FLAG` i.e. it will only work if we set the dook on initialization

Could be resolved by setting the fee as `DYNAMIC_FEE_FLAG` regardless of whether we're setting a dook during initialization, although I still need to consider whether there are any side effects to doing this